File: 515-largest-value-in-tree-row/go/main.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func largestValues(root *TreeNode) []int {
	q := []*TreeNode{}
	res := []int{}

	if root == nil {
		return res
	}

	q = append(q, root)

	for len(q) != 0 {
		temp := []*TreeNode{}

		for len(q) != 0 {
			temp = append(temp, q[0])
			q = q[1:]
		}

		maxVal := temp[0].Val

		for _, node := range temp {
			if node.Val > maxVal {
				maxVal = node.Val
			}
			if node.Left != nil {
				q = append(q, node.Left)
			}

			if node.Right != nil {
				q = append(q, node.Right)
			}
		}

		res = append(res, maxVal)
	}

	return res
}

File: 515-largest-value-in-tree-row/cpp/main.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
 public:
  vector<int> largestValues(TreeNode* root) {
    vector<int> res;
    queue<TreeNode*> q;  // bfs
    if (root == NULL) {
      return res;
    }
    q.push(root);

    while (!q.empty()) {
      vector<TreeNode*> level;
      while (!q.empty()) {
        level.push_back(q.front());
        q.pop();
      }
      int maxVal = level[0]->val;
      for (auto i : level) {
        maxVal = max(maxVal, i->val);
        if (i->left != NULL) {
          q.push(i->left);
        }
        if (i->right != NULL) {
          q.push(i->right);
        }
      }
      res.push_back(maxVal);
    }

    return res;
  }
};
